src/js/_enqueues/lib/emoji-loader.js

```diff
@@ -57,25 +57,38 @@
 	}
 
 	/**
-	 * Get supports from session.
+	 * @typedef SessionSupportTests
+	 * @type {object}
+	 * @property {number} timestamp
+	 * @property {SupportTests} supportTests
+	 */
+
+	/**
+	 * Get support tests from session.
 	 *
 	 * @since 6.3.0
 	 *
 	 * @private
 	 *
-	 * @returns {?SupportTests} Supports or null if not set.
+	 * @returns {?SupportTests} Support tests, or null if not set or older than 1 day.
 	 */
-	function getSessionSupports() {
+	function getSessionSupportTests() {
 		if (
 			typeof sessionStorage !== 'undefined' &&
 			sessionStorageKey in sessionStorage
 		) {
 			try {
-				var supports = JSON.parse(
+				/** @type {SessionSupportTests} */
+				var item = JSON.parse(
 					sessionStorage.getItem( sessionStorageKey )
 				);
-				if ( typeof supports === 'object' ) {
-					return supports;
+				if (
+					typeof item === 'object' &&
+					typeof item.timestamp === 'number' &&
+					new Date().valueOf() < item.timestamp + 86400 &&
```

Comment:
this is ultimately tied to the version of Twemoji that WordPress ships isn't it? I think that the fallback case we're discussing is somewhat benign: if someone reboots and re-opens tabs and now their OS adds additional support for Unicode 14 Emoji, the worst thing that happens is that we continue to replace them until those tabs or the session storage clears. That's not that bad.

With this change we are going back to computing this at least once per day, which for many sites could very well be every page load.

Is the risky scenario that we update Twemoji in Core and can handle the next set of Emoji additions, but because this script cached the results it doesn't replace them and then the page is rendered improperly?

In that case could we build in some sentinel indicating the Twemoji version and then handle this deterministically? Update automatically when Twemoji updates or in the natural flow of browser reloads? Otherwise be okay with being a bit too cautious for some amount of time after Unicode releases?

Reply:
> we should check to see if all of the tests are present?

actually this would be way better than what I suggested and wouldn't need to have any coupling with Twemoji. that's a brilliant idea, because the tests are trivial to serialize.

it wouldn't just be the names, but rather the specific strings we compare, or at least one of the two in comparison.

the ultimate goal is to use some canary string that we know is only supported in some version of the Unicode release. if that canary isn't present then we need Twemoji to fill in the gap.

if we don't add the test for it, the version of Twemoji wouldn't matter because we wouldn't be queuing it up anyway. supposing that Unicode releases a new version of the Emoji and a new character is available, we would need to add a new test to check if it renders as we expect. until then, the tests here would remain the same even though the rendering might be broken.

Reply:
If the session storage cache is varied by the strings we compare, I don't think we can easily do so without computing a hash of `browserSupportsEmoji` and using this string for comparison. Perhaps a more efficient means would be to just introduce a `emojiLoaderVersion` that we bump whenever we touch the `browserSupportsEmoji` function to modify the strings?

Reply:
Well, the cache should be invalidated whenever test strings change (when `browserSupportsEmoji()` is changed) _or_ the OS emoji support changes. I think going ahead with just the 1 day expiration for now is adequate. We can consider further optimization after beta1.

Reply:
sure, my point was the "OS emoji support changes" doesn't necessitate an update of the cache. it's an edge case whose failure mode is that we're over-cautious vs. rendering a broken page.

it hit me that _for future work_ we could revamp this entire function to focus on versions of the Unicode spec to support instead of randomly picking strings we happen to have seen fail.

if we see support for Unicode 12 we never have to check again.

also, if we do this, maybe it's possible with Twemoji (I don't know) to specify through the URL to only load in support for versions newer than X. This could make the loader even more useful by keeping all browsers updated as Unicode revises itself while avoiding the full load of the Emoji database Twemoji can support.

> If the session storage cache is varied by the strings we compare, I don't think we can easily do so without computing a hash of the browserSupportsEmoji function and using this string for comparison.

I don't follow why, but that's fine; can skip that. Using the tests as you suggested is probably good enough for now. A version number seems good enough too. Again, my point was that caching for 24 hours is likely _almost_ identical to not caching, and wildly too conservative. It's running this script more often (against the spirit of this PR) in order to catch a failure mode that isn't a real failure.

Don't really think it's a major point worth blocking this PR, but I do find it possibly a bit silly to add in, and not _better_ than having no cache, or a cache on a week or a month basis.

Reply:
I did so in b27e7130ded54964ece4304d92b9e71ba84b7f2a.

Reply:
> So you'd be happier with a week cache?

I'm happy with all you've done in this PR no matter how you choose the timeout. 😄

Reply:
@westonruter @dmsnell Sorry I didn't do a better job of explaining my thinking.

The thought behind the comment was indeed related to the twemoji (or WP version). I was thinking of the following situation:

* A user opens the browser, sessionStorage is set
* The user leaves the browser open overnight
* The site admins update WP to a version that includes an update to twemoji and the related sniffs 
* The user returns and continues in the existing tab, as they navigate WP scripts will use the new version
* Without an expiration control the sniff would continue to be based on the previous tests.

Dennis, you raise a good point that the stale check could be based on the twemoji or WP version. That would allow the sessionStorage to remain valid until a potential change is detected

The origin of this thought was when I worked for a news organisation and discovered that journos keep their browsers open forever, constantly duplicating tabs. The team I was on was using both session and local storage and we had to add some code to invalidate it upon a deploy. Several years later, this experience continues to be traumatic. :)

Reply:
> Several years later, this experience continues to be traumatic. :)

That's understandable. Also why I like finding ways to make durable cache keys (for instance, "properly renders Emoji added in Unicode 12"

Been burnt the other way too, with "we'll only leave in daily expiration until the project stablizes…five years later" 😆 

It's like cache invalidation is hard or something.

Honestly I'd love to look into redoing this after @westonruter suggested storing the test names as the key. Twemoji incorporates what looks like a 13 KB regular expression pattern that includes the entire database of replacements up through Unicode 14 (curious if and when they update to Unicode 15). If these were separated out we could potentially cut out 10 KB+ of download when required, and our session cache would remain durable under all but the most extreme edge cases (someone downgrades their operating system and re-opens a browser on the same tabs and is also not okay with missing Emoji)

Reply:
I've opened a Trac ticket we can use to track this going forward: Core-58663. I've marked it very low priority.